src/ja/server/scheduler/scheduler.py

```diff
@@ -11,20 +16,64 @@ class Scheduler:
     jobs which are still runnable.
     """
 
-    def __init__(self, algorithm: SchedulingAlgorithm):
+    def __init__(self, algorithm: SchedulingAlgorithm, dispatcher: Dispatcher, special_resources: Dict[str, int]):
         """!
         Initialize a Scheduler.
 
-        @param algorithm The scheduling algorithm to use to assign jobs to
-          work machines.
+        @param algorithm The scheduling algorithm to use to assign jobs to work machines.
+        @param dispatcher The dispatcher to use for sending commands to the work machines.
+        @param special_resources The available special resources.
         """
+        self._algorithm = algorithm
+        self._dispatcher = dispatcher
+        self._special_resources = deepcopy(special_resources)
+        self._total_special_resources = deepcopy(special_resources)
+
+    @property
+    def special_resources(self) -> Dict[str, int]:
+        """!
+        The amount of free special resources.
+        """
+        return self._special_resources
+
+    @property
+    def total_special_resources(self) -> Dict[str, int]:
+        """!
+        The total amount of special resources.
+        """
+        return self._total_special_resources
 
     def reschedule(self, database: ServerDatabase) -> None:
         """!
-        Fetches the current schedule from the database, unassigns work machines
-        from jobs which are no longer runnable, and redistributes the other
-        jobs using the scheduling algorithm. Finally, the each modified job and
-        work machine are updated in the database.
+        Fetches the current schedule from the database and redistributes the jobs using the scheduling algorithm.
+        Then each modified job and work machine are updated in the database.
 
         @param database The database to fetch job schedule from.
         """
+        available_machines = list(filter(lambda m: m.state == WorkMachineState.ONLINE, database.get_work_machines()))
+        new_schedule = self._algorithm.reschedule_jobs(database.get_current_schedule(),
+                                                       available_machines, self.special_resources)
+
+        # Reset resources and recalculate them
+        for machine in available_machines:
+            machine.resources.deallocate(machine.resources.total_resources - machine.resources.free_resources)
+        self._special_resources = deepcopy(self._total_special_resources)
+
+        for job in new_schedule:
```

Comment:
isn't it going to be better to call this job -> job_entry, so that you don't have job.job.

Reply:
this minor things, I will resolve it.